### ISSUE TITLE
Accuracy of yolov4-tiny not improving even after trying suggestions in the repo

### ISSUE THREAD

**Issue by poornimajd** [AlexeyAB/darknet#7739]
Hello,
 I have 8 classes,around 10k train images,2k validation images.I have tried including some **negative images,reduced learning rate and made sure that the resolution of train and test images matches,trained it for 16000 iterations.I have also tried to increase the training image resolution in the cfg file to 640x640.**
But even after trying all these ,also **trying to add extra data**,the map scores on the test data do not increase at all ,it hardly increases by 1 percent or something.Why is this happening?Can any one please give any suggestion on this?

The **train** **mAP is around 72-73% but testing mAP is close to 61%.**

Screenshot while training-
![chart](https://user-images.githubusercontent.com/31023599/119260798-0d9ffa80-bbf2-11eb-80ae-b2b2e878c58d.png)

Thank you

**Comment by woosal1337** [comment]
Are you sure of the perfect or at least near to perfect classification of your data?

**Comment by poornimajd** [comment]
@woosal1337 Thanks for the reply.


> Are you sure of the perfect or at least near to perfect classification of your data?

Do you mean if the labels are correct?
Yes they are.I checked the data.

**Comment by woosal1337** [comment]
Yes, I do actually mean the labels. This is mainly also caused by the data quality actually. Make sure instead of having more data, to have much more high-quality data to train.

**Comment by poornimajd** [comment]
Thanks for the suggestion.Yeah this is true,but I added some 400 images similar to that of test data too,but even then accuracy is not greatly improved.I am not sure if this is the limitation of yolov4-tiny?

**Comment by woosal1337** [comment]
That is actually interesting. However, do not also forget that you are using the "tiny" model. As far as I remember expecting the same result as the default YOLO model would not also be reasonable on the "tiny". Please, correct me if I am mistaken.

**Comment by poornimajd** [comment]
This is true,but when I had trained yolov4, I had not even given data similar to test set and it worked quite well,better than this tiny yolo,which is having data similar to test set,and hence I had some higher expectations.

**Comment by bulatnv** [comment]
Hello, @poornimajd.

Can you show some of your test, train samples?

**Comment by poornimajd** [comment]
@bulatnv thanks for the reply.
Some of the train and test images are as follows-
https://github.com/poornimajd/newdemo
Train folder has training images and test folder has test images

There are 8 classes-person,animal,motorcycle,bicycle,autorickshaw,car,truck,bus

counters_per_class-19424,354,14008,1336,4448,12643,4576,3016
(in the order as I have written the names)

anchors= 24,58,59,80,38,147,91,209,182,302,367,498

I think I have provided the required info. Also I agree there is class imbalance,but yolo does not really have this issue right?
Also leaving the class animal(having least number of samples),the scores for other classes still do not improve much.

**Comment by bulatnv** [comment]
@poornimajd 
Have you tried to change anchors in cfg?

in your case:
```
218) mask = 3,4,5
219) anchors = 91,209, 182,302, 367,498
...
267) mask = 0,1,2
268) anchors = 24,58, 59,80, 38,147
```
Read more about it here: https://github.com/AlexeyAB/darknet/issues/6740

> Also leaving the class animal(having least number of samples),the scores for other classes still do not improve much.

You mean that animal class is not the worst?

**Comment by poornimajd** [comment]
Thanks for the reply!
> Have you tried to change anchors in cfg?

No, I have not tried it.Actually I am not really an expert in this,so did not touch this part.
Does recalculating anchors improve results significantly?



> You mean that animal class is not the worst?

The animal class is the worst,and also has the least number of samples for training,hence the low accuracy in animal class is acceptable but I am not sure about other classes.

**Comment by bulatnv** [comment]
> Does recalculating anchors improve results significantly?

It will, but not very much.

> The animal class is the worst

Try to oversample it. Use random affine transformations (rotations, scale) + random horizontal flips, random crops.
Try to increase number of animal objects at least to 3000. Same with motorcicles.
After that recalculate class counters and add in the yolo layer:

`counters_per_class = 19424,354,14008,1336,4448,12643,4576,3016`

**Comment by poornimajd** [comment]
Thanks for the suggestion,will try it out ,hope the accuracy improves.

**Comment by Micky-123** [comment]
@poornimajd : were you able to improve the accuracy eventually?

**Comment by Micky-123** [comment]
@poornimajd : I am also facing similar issue.. Can you please let me know if you were able to improve the accuracy..

**Comment by stephanecharette** [comment]
> Can you please let me know if you were able to improve the accuracy

See the "if the results aren't great" section here:  https://www.ccoderun.ca/programming/darknet_faq/#tldr

**Comment by poornimajd** [comment]
@Micky-123 ,I have to collect more data ,which might improve the accuracy but I have not been able to work on it much for now.

**Comment by Micky-123** [comment]
> @Micky-123 ,I have to collect more data ,which might improve the accuracy but I have not been able to work on it much for now.

Thanks! If you have interesting finding, do update... it will help others who are in the same situation.
